src/core/server/plugins/plugins_system.ts

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { KibanaCore } from '../../types';
+import { Logger } from '../logging';
+import { Plugin, PluginName } from './plugin';
+import { createPluginInitializerCore, createPluginStartCore } from './plugins_core';
+
+/** @internal */
+export class PluginsSystem {
```

Comment:
I'm OK with this naming choice for the sake of an initial merge, but I think in the end we should change this name. In isolation I think `PluginsSystem` is fine, but in the spirit of having the stuff in `core` be an example for future development within core and plugins alike, I think a more clear structure is in order for the the plugin module.

Let's look to core itself for inspiration here.

There's a top level service definition for core, we happen to call it `Server`, but effectively it is just the same thing as any other service definition. It's a class that takes in some constructor arguments from its initializer, it contains start/stop lifecycle events to kick off and tear down the relevant domain logic, etc.

Simply looking at the top-most directory structure of the core module you get a pretty good idea of its features, responsibilities, etc.

```
core
  config
  http
  legacy
  logging
  plugins
  index.js // the interface for the core module
```

The directory structure can't describe _how_ these features and responsibilities work together or how they get exposed, but it at least gives a good indication of _what_ the core module is all about.

There are currently also some oddities in core that I didn't list above like `root` and `dev`, but I think of those things as technical debt that we chose to take on for the sake of practicality and that we can resolve later. And I left out some name and path details for the sake of explanation.

We should strive to have that same level of clarity from each of the services within core as well.

When I crack open plugin, I'd expect to see a file structure that was equally as expressive as that of core itself. Perhaps something like:

```
core
  plugins
    config    // config schema validation
    discovery // static manifest discovery
    loader    // importing plugin interface and instantiation
    lifecycle // invoke start/stop/init(?) with relevant contracts
    index.js  // the interface for the core/plugins module
```

I might not have the right break down there, but hopefully it illustrates my point.

This feedback is really not unique to plugins, so if you'd prefer not to heed it in this PR, I'm fine with that. But let's at least get on the same page about it and then we can apply it to future service design and come back and refactor some of the existing core service designs at a later date.

Reply:
On further reflection, this PR is already pretty massive, so I don't think overhauling the existing shape of the plugin module in it makes sense. I'd love your feedback so we can get aligned on it going forward.

Reply:
Hehe, I was struggling with finding the right name and just ended up with what we had in new-platform branch (same as I did for `Server` and `Root`) :) I don't like our `Server`, `Root`, `PluginSystem` and a couple more things, these are confusing and meaningless names/notions IMO.

Up until this point `SomethingService` was a central thing for `core/something` and `SomethingModule` served as a grouping mechanism, mainly because of simplicity of `Something`, but I think `ModuleSomething` (e.g. `core/something/index.ts` in your example) should become a central thing and be an "interface" for `Something` that other `SomethingElse` from the core can interact with. Details to be discussed though.

> When I crack open plugin, I'd expect to see a file structure that was equally as expressive as that of core itself.
> On further reflection, this PR is already pretty massive, so I don't think overhauling the existing shape of the plugin module in it makes sense. I'd love your feedback so we can get aligned on it going forward.

I haven't dug that deep yet, but I think that's an interesting point that totally makes sense to me. And I agree that we can handle this in follow-ups more effectively.